### PR TITLE
[PM-5275] Migrate state in Fido2ClientService to State Providers

### PR DESCRIPTION
**Pull request by gbubemismith**
## Type of change

```
- [ ] Bug fix
- [ ] New feature development
- [X] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective
Migrate enablePasskeys state to state provider.

## Code changes

- **libs/common/src/vault/services/vault-settings/vault-settings.service.ts:** Created vault settings service to hold settings related to the vault.
- **libs/common/src/platform/state/state-definitions.ts:** Added state definition to store state on disk.
- **libs/common/src/vault/services/key-state/enable-passkey.state.ts:** Created the `enablePasskeys` key definition.
- **libs/common/src/state-migrations/migrations/14-move-enable-passkeys-to-state-providers.ts:** Migrates enablePasskeys from the `global` state.
- **libs/common/src/platform/services/state.service.ts:** Removed get and set enable passkeys from the state service.
- **apps/browser/src/popup/settings/options.component.ts:** Updated to use vault settings service.

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

### CODE REVIEW

**Review by shane-melton — CHANGES_REQUESTED** [review]
Awesome work! Thanks for going ahead and creating the `VaultSettingsService` and getting that started for us. 

I have a couple questions/suggestions, let me know that you think!

File: libs/common/src/platform/state/state-definitions.ts [hunk 1]
```diff
@@ -26,3 +26,5 @@ export const ENVIRONMENT_DISK = new StateDefinition("environment", "disk");
 
 export const GENERATOR_DISK = new StateDefinition("generator", "disk");
 export const GENERATOR_MEMORY = new StateDefinition("generator", "memory");
+
+export const FIDO2_DISK = new StateDefinition("fido2", "disk");
```
[the remark this answers is not in the document]
Comment (gbubemismith) [reply libs/common/src/platform/state/state-definitions.ts:30]: good point

File: apps/browser/src/background/main.background.ts
[the remark this answers is not in the document]
Comment (gbubemismith) [reply apps/browser/src/background/main.background.ts:1]: I think this should be done in the vault settings creation ticket

File: libs/common/src/platform/state/state-definitions.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (gbubemismith) [reply libs/common/src/platform/state/state-definitions.ts]: Addressed f0c4e2be0d51ea105272219a09e9a7297b767091

File: libs/common/src/vault/services/vault-settings/vault-settings.service.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+import { USER_ENABLE_PASSKEYS } from "../key-state/enable-passkey.state";
+
+/**
+ * {@link VaultSettingsServiceAbstraction}
+ */
+export class VaultSettingsService implements VaultSettingsServiceAbstraction {
+  private enablePasskeysState: GlobalState<boolean>;
+
+  constructor(private stateProvider: StateProvider) {
+    this.enablePasskeysState = stateProvider.getGlobal(USER_ENABLE_PASSKEYS);
```
[the remark this answers is not in the document]
Comment (gbubemismith) [reply libs/common/src/vault/services/vault-settings/vault-settings.service.ts]: Addressed here f0c4e2be0d51ea105272219a09e9a7297b767091

File: libs/common/src/vault/services/vault-settings/vault-settings.service.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  }
+
+  /**
+   * {@link VaultSettingsServiceAbstraction.enablePasskeys$}
+   */
+  get enablePasskeys$(): Observable<boolean> {
```
[the remark this answers is not in the document]
Comment (gbubemismith) [reply libs/common/src/vault/services/vault-settings/vault-settings.service.ts]: Addressed here f0c4e2be0d51ea105272219a09e9a7297b767091

File: libs/common/src/state-migrations/migrations/14-move-enable-passkeys-to-state-providers.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+export class EnablePasskeysMigrator extends Migrator<13, 14> {
+  async migrate(helper: MigrationHelper): Promise<void> {
+    const global = await helper.get<ExpectedGlobalType>("global");
+
+    if (global?.enablePasskeys != null) {
+      await helper.setToGlobal(USER_ENABLE_PASSKEYS, global.enablePasskeys);
+    }
+
+    delete global?.enablePasskeys;
+    await helper.set("global", global);
```
[the remark this answers is not in the document]
Comment (gbubemismith) [reply libs/common/src/state-migrations/migrations/14-move-enable-passkeys-to-state-providers.ts]: Addressed here f0c4e2be0d51ea105272219a09e9a7297b767091

**Review by shane-melton — CHANGES_REQUESTED** [review]
Looks great! Thanks for making those changes I think it turned out well!

Only thing I noticed is we should still add `readonly` to `enablePasskeys$` in the abstraction class as well as the implementation so its exposed to consumers of the service.

File: libs/common/src/vault/abstractions/vault-settings/vault-settings.service.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
+import { Observable } from "rxjs";
+/**
+ * Service for managing vault settings.
+ */
+export abstract class VaultSettingsService {
+  /**
+   * An observable monitoring the state of the enable passkeys setting.
+   * The observable updates when the setting changes.
+   */
+  enablePasskeys$: Observable<boolean>;
```
[the remark this answers is not in the document]
Comment (gbubemismith) [reply libs/common/src/vault/abstractions/vault-settings/vault-settings.service.ts:10]: Thanks I missed that

File: libs/common/src/vault/services/fido2/fido2-client.service.spec.ts [hunk 6]
```diff
@@ -40,10 +42,17 @@ describe("FidoAuthenticatorService", () => {
...
+    client = new Fido2ClientService(
+      authenticator,
+      configService,
+      authService,
+      stateService,
+      vaultSettingsService,
+    );
     configService.serverConfig$ = of({ environment: { vault: VaultUrl } } as any);
-    stateService.getEnablePasskeys.mockResolvedValue(true);
+    vaultSettingsService.enablePasskeys$ = of(true);
```
Comment (shane-melton) [inline libs/common/src/vault/services/fido2/fido2-client.service.spec.ts:55]: It looks like the tests are now failing due to assigning of the `readonly` value. I think we should be able to replace this line with: `jest.spyOn(vaultSettingsService, "enablePasskeys$", "get").mockReturnValue(of(true));`

**Review by shane-melton — CHANGES_REQUESTED** [review]
Looks like we just need to get the tests working again and then we're good to go.

**Review by shane-melton — DISMISSED** [review]

File: libs/common/src/vault/services/vault-settings/vault-settings.service.ts [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+import { GlobalState, StateProvider } from "../../../platform/state";
+import { VaultSettingsService as VaultSettingsServiceAbstraction } from "../../abstractions/vault-settings/vault-settings.service";
+import { USER_ENABLE_PASSKEYS } from "../key-state/enable-passkey.state";
+
+/**
+ * {@link VaultSettingsServiceAbstraction}
+ */
+export class VaultSettingsService implements VaultSettingsServiceAbstraction {
+  private enablePasskeysState: GlobalState<boolean> =
+    this.stateProvider.getGlobal(USER_ENABLE_PASSKEYS);
```
Comment (justindbaur) [inline libs/common/src/vault/services/vault-settings/vault-settings.service.ts:12]: Kinda loving this style of initializing y'all are doing. Might update some documentation to show this off more prominently.

File: libs/common/src/state-migrations/migrations/14-move-enable-passkeys-to-state-providers.ts [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    const global = await helper.get<ExpectedGlobalType>("global");
+
+    if (global) {
+      if (global.enablePasskeys != null) {
+        await helper.setToGlobal(USER_ENABLE_PASSKEYS, global.enablePasskeys);
+      }
+
+      delete global?.enablePasskeys;
+      await helper.set("global", global);
+    }
```
Comment (justindbaur) [inline libs/common/src/state-migrations/migrations/14-move-enable-passkeys-to-state-providers.ts:26]: I prefer the style of migration that you did in folder service. Basically chain `?.` on the value you care about. If it's not-null, migrate it and avoid re-updating any state if you didn't do any migration. https://github.com/bitwarden/clients/pull/7682/files#diff-571eee65ea4a963fe2b8f96fa783296f55c3111da94db0459b11d58697a4b107R29-R34

**Review by justindbaur — CHANGES_REQUESTED** [review]
One minor request on the migration, but looks great!

**Review by justindbaur — DISMISSED** [review]

**Review by shane-melton — CHANGES_REQUESTED** [review]
Looks like there are some merge conflicts and the `CURRENT_VERSION` in `migrate.ts` has incremented again on `main`.

**Review by shane-melton — APPROVED** [review]

**Review by justindbaur — DISMISSED** [review]

**Review by shane-melton — APPROVED** [review]

**Review by justindbaur — APPROVED** [review]

**Pull request merged**